### ISSUE TITLE
Discrepancy Between Reported and Reproduced Results for Surgformer

### ISSUE THREAD

**Issue by Hyeongyuc96hutom** [isyangshu/Surgformer#10]
Dear authors,

I greatly appreciate your work on Surgformer, which has achieved state-of-the-art performance in surgical phase recognition. I have followed the instructions provided in the README and conducted both training and evaluation without any modifications to the code.

However, the results I obtained are lower than those reported in your paper. I would like to verify whether I have correctly followed the intended training and evaluation procedures. To facilitate this, I am sharing the scripts I used for training and testing.

For the pretrained model, I used the [TimeSformer pretrained weight](https://www.dropbox.com/scl/fi/zcn6byf10i4r0hhojjten/TimeSformer_divST_8x32_224_K400.pyth?rlkey=azfkkmb0qalhgt9vxofhwje54&e=1&dl=0) as specified in the README. Additionally, I am attaching the inference results merged using convert_cholec80.py, including both phase annotations and predictions. [surgformer_results.zip](https://github.com/user-attachments/files/19310136/surgformer_results.zip)

### 1. With --cut_black enabled
Training Command:
```shell
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 python -m torch.distributed.launch \
--nproc_per_node=8 \
--master_port 12324 \
~/Surgformer/downstream_phase/run_phase_training.py \
--batch_size 10 \
--epochs 50 \
--save_ckpt_freq 10 \
--model surgformer_HTA_KCA \
--pretrained_path ~/Surgformer/pretrained_params/TimeSformer_divST_8x32_224_K400.pyth \
--mixup 0.8 \
--cutmix 1.0 \
--smoothing 0.1 \
--lr 5e-4 \
--layer_decay 0.75 \
--warmup_epochs 5 \
--data_path /raid/dataset/public/cholec80_surgformer \
--eval_data_path /raid/dataset/public/cholec80_surgformer \
--nb_classes 7 \
--data_strategy online \
--output_mode key_frame \
--num_frames 16 \
--sampling_rate 4 \
--data_set Cholec80 \
--data_fps 1fps \
--output_dir ~/surgformer/experiments-cut_black \
--log_dir ~/surgformer/experiments-cut_black \
--num_workers 10 \
--dist_eval \
--enable_deepspeed \
--no_auto_resume \
--cut_black
```

Testing Command (Best epoch: 22):
```shell
CUDA_VISIBLE_DEVICES=2,3 python -m torch.distributed.launch \
--nproc_per_node=2 \
--master_port 12330 \
~/Surgformer/downstream_phase/run_phase_training.py \
--batch_size 32 \
--epochs 50 \
--model surgformer_HTA_KCA \
--pretrained_path ~/Surgformer/pretrained_params/TimeSformer_divST_8x32_224_K400.pyth \
--mixup 0.8 \
--cutmix 1.0 \
--smoothing 0.1 \
--lr 5e-4 \
--layer_decay 0.75 \
--warmup_epochs 5 \
--data_path /raid/dataset/public/cholec80_surgformer \
--eval_data_path /raid/dataset/public/cholec80_surgformer \
--nb_classes 7 \
--data_strategy online \
--output_mode key_frame \
--num_frames 16 \
--sampling_rate 4 \
--eval \
--finetune ~/surgformer/experiments-cut_black/surgformer_HTA_KCA_Cholec80_0.0005_0.75_online_key_frame_frame16_Fixed_Stride_4/checkpoint-best/mp_rank_00_model_states.pt \
--data_set Cholec80 \
--data_fps 1fps \
--output_dir ~/surgformer/experiments-cut_black/results \
--log_dir ~/surgformer/experiments-cut_black/results \
--num_workers 10 \
--dist_eval \
--enable_deepspeed \
--no_auto_resume \
--cut_black
```

Evaluation Results:
```shell
% Relaxed 10 sec
================================================
                    Phase|  Jacc|  Prec|   Rec|
================================================
              Preparation| 70.20| 95.30| 73.84|
---------------------------------------------
  CalotTriangleDissection| 90.14| 92.64| 97.31|
---------------------------------------------
          ClippingCutting| 87.36| 95.17| 92.29|
---------------------------------------------
    GallbladderDissection| 87.96| 93.69| 94.52|
---------------------------------------------
     GallbladderPackaging| 75.10| 85.96| 88.71|
---------------------------------------------
      CleaningCoagulation| 71.09| 82.25| 87.04|
---------------------------------------------
    GallbladderRetraction| 79.52| 90.20| 90.36|
---------------------------------------------
================================================
Mean jaccard: 80.19+- 8.37
Mean f1-score: 88.69+- 6.59
Mean accuracy: 91.78+- 5.73
Mean precision: 90.75+- 4.97
Mean recall: 89.15+- 7.59

% UnRelaxed
================================================
                    Phase|  Jacc|  Prec|   Rec|
================================================
              Preparation| 68.60| 94.48| 72.96|
---------------------------------------------
  CalotTriangleDissection| 89.75| 92.27| 97.10|
---------------------------------------------
          ClippingCutting| 81.93| 90.85| 89.62|
---------------------------------------------
    GallbladderDissection| 86.84| 92.70| 93.56|
---------------------------------------------
     GallbladderPackaging| 71.12| 82.73| 84.71|
---------------------------------------------
      CleaningCoagulation| 65.81| 74.19| 83.04|
---------------------------------------------
    GallbladderRetraction| 71.44| 85.18| 81.45|
---------------------------------------------
================================================
Mean jaccard: 76.50+- 9.52
Mean f1-score: 85.13+- 6.95
Mean accuracy: 90.92+- 5.63
Mean precision: 87.49+- 7.25
Mean recall: 86.06+- 8.11

```

### 2. Without --cut_black
Training Command:
```shell
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 python -m torch.distributed.launch \
--nproc_per_node=8 \
--master_port 12324 \
~/Surgformer/downstream_phase/run_phase_training.py \
--batch_size 10 \
--epochs 50 \
--save_ckpt_freq 10 \
--model surgformer_HTA_KCA \
--pretrained_path ~/Surgformer/pretrained_params/TimeSformer_divST_8x32_224_K400.pyth \
--mixup 0.8 \
--cutmix 1.0 \
--smoothing 0.1 \
--lr 5e-4 \
--layer_decay 0.75 \
--warmup_epochs 5 \
--data_path /raid/dataset/public/cholec80_surgformer \
--eval_data_path /raid/dataset/public/cholec80_surgformer \
--nb_classes 7 \
--data_strategy online \
--output_mode key_frame \
--num_frames 16 \
--sampling_rate 4 \
--data_set Cholec80 \
--data_fps 1fps \
--output_dir ~/surgformer/experiments \
--log_dir ~/surgformer/experiments \
--num_workers 10 \
--dist_eval \
--enable_deepspeed \
--no_auto_resume
```

Testing Command (Best epoch: 19):
```shell
CUDA_VISIBLE_DEVICES=2,3 python -m torch.distributed.launch \
--nproc_per_node=2 \
--master_port 12330 \
~/Surgformer/downstream_phase/run_phase_training.py \
--batch_size 32 \
--epochs 50 \
--model surgformer_HTA_KCA \
--pretrained_path ~/Surgformer/pretrained_params/TimeSformer_divST_8x32_224_K400.pyth \
--mixup 0.8 \
--cutmix 1.0 \
--smoothing 0.1 \
--lr 5e-4 \
--layer_decay 0.75 \
--warmup_epochs 5 \
--data_path /raid/dataset/public/cholec80_surgformer \
--eval_data_path /raid/dataset/public/cholec80_surgformer \
--nb_classes 7 \
--data_strategy online \
--output_mode key_frame \
--num_frames 16 \
--sampling_rate 4 \
--eval \
--finetune ~/surgformer/experiments/surgformer_HTA_KCA_Cholec80_0.0005_0.75_online_key_frame_frame16_Fixed_Stride_4/checkpoint-best/mp_rank_00_model_states.pt \
--data_set Cholec80 \
--data_fps 1fps \
--output_dir ~/surgformer/experiments/results \
--log_dir ~/surgformer/experiments/results \
--num_workers 10 \
--dist_eval \
--enable_deepspeed \
--no_auto_resume \
--cut_black
```

Evaluation Results:
```shell
% Relaxed 10 sec
================================================
                    Phase|  Jacc|  Prec|   Rec|
================================================
              Preparation| 71.04| 97.08| 73.68|
---------------------------------------------
  CalotTriangleDissection| 89.36| 92.40| 96.43|
---------------------------------------------
          ClippingCutting| 83.93| 89.24| 93.89|
---------------------------------------------
    GallbladderDissection| 88.61| 94.87| 94.08|
---------------------------------------------
     GallbladderPackaging| 78.13| 85.78| 92.28|
---------------------------------------------
      CleaningCoagulation| 68.76| 84.39| 80.32|
---------------------------------------------
    GallbladderRetraction| 74.00| 80.96| 94.71|
---------------------------------------------
================================================
Mean jaccard: 79.12+- 8.35
Mean f1-score: 88.22+- 7.14
Mean accuracy: 91.66+- 5.37
Mean precision: 89.25+- 5.88
Mean recall: 89.34+- 8.73

% UnRelaxed
================================================
                    Phase|  Jacc|  Prec|   Rec|
================================================
              Preparation| 68.92| 96.41| 72.17|
---------------------------------------------
  CalotTriangleDissection| 89.03| 92.09| 96.26|
---------------------------------------------
          ClippingCutting| 77.61| 83.41| 92.40|
---------------------------------------------
    GallbladderDissection| 86.99| 93.71| 92.53|
---------------------------------------------
     GallbladderPackaging| 74.03| 82.29| 88.93|
---------------------------------------------
      CleaningCoagulation| 64.81| 80.57| 76.66|
---------------------------------------------
    GallbladderRetraction| 69.49| 77.11| 89.52|
---------------------------------------------
================================================
Mean jaccard: 75.84+- 9.26
Mean f1-score: 84.76+- 7.44
Mean accuracy: 90.80+- 5.27
Mean precision: 86.51+- 7.44
Mean recall: 86.92+- 8.96
```

Given the discrepancy in performance, I would appreciate any insights you can provide regarding potential reasons for the difference. Is there anything I might have overlooked in my setup?

Thank you for your time and assistance. I look forward to your response.

Best regards

**Comment by isyangshu** [comment]
Well, thanks.
Actually, we use 24*4 for evaluation on Cholec80 dataset.
Could you try this setting for evaluation? And I look forward to your subsequent results.

<img width="767" alt="Image" src="https://github.com/user-attachments/assets/913532aa-0080-48cf-9219-4c6343486267" />

**Comment by Hyeongyuc96hutom** [comment]
Thank you for your quick response. I see that I used num_frames=16, whereas the experiments in the paper were conducted with num_frames=24.

**Comment by isyangshu** [comment]
However, I notice that your results are still lower than our performance on 16*4.
90.9 -> 91.7 (~0.8)
91.7 -> 92.5 (~0.8)
But I don't know where the difference comes from. From my own experiments, larger batches do not lead to better performance (I noticed that you used 8 cards to speed up the experiments). Meanwhile, if you need, I can share my parameters with you to determine if the difference in performance is related to the preparation of the dataset.

<img width="1047" alt="Image" src="https://github.com/user-attachments/assets/8d1296f1-80fc-42d6-97cb-2537b502f9d2" />

**Comment by Hyeongyuc96hutom** [comment]
I used 8 GPUs for faster training.
From a DDP perspective, to keep the model weight update iterations as close to yours as possible, I used a batch size of 10 per GPU (80 datasets per iteration) with 8 GPUs. (For reference, your setting was 72 datasets per iteration since you used 3 GPUs with a batch size of 24.)

Of course, from a DDP perspective, since the gradients of each worker are averaged based on the number of num_workers, this could have an impact on performance.

Thank you for sharing that using a larger batch size did not lead to better performance.
To check if the difference comes from the dataset preparation, would you be able to share your weights with me?

----
And I have one more question while reading your paper.
You mentioned that the frame rate is set to R=4 in the paper. Does this mean that the video data was subsampled to 4 fps?
<img width="993" alt="Image" src="https://github.com/user-attachments/assets/dfbf6d24-66ee-48e4-a554-44e024544304" />

I found the following setting in the training script:

```shell
--sampling_rate 4 \
--data_set Cholec80 \
--data_fps 1fps \
```

From my understanding of the code, it seems that the data preparation process uses subsampled data at 1 fps. Is this correct? Could you clarify what sample_rate 4 means in this context?

**Comment by isyangshu** [comment]
The sample_rate 4 means the sampling interval for input, like [0, 4, 8, 16], [1, 5, 9, 13]. You can try to print the frame list for deep understanding. Therefore, we always use 1fps data for training and testing.

You can find the parameters here: https://hkustconnect-my.sharepoint.com/:f:/g/personal/syangcw_connect_ust_hk/EvR50R8_O5NOj24xSCRqBisBqMBx3ZzWlJyjIv8S_DseJA?e=LOOxVq

**Comment by carlosh93** [comment]
Hi @isyangshu I have the same issue when trying to reproduce the training. I double-checked the model you shared, and it is showing the same accuracy you reported in the paper. However, I would like to know why I cannot reproduce the training. Could you please share your training script? I'm using 2 GPUS, A100, does this script make sense?

```
CUDA_VISIBLE_DEVICES=0,1 python -m torch.distributed.launch \
--nproc_per_node=2 \
--master_addr "localhost" \
--master_port 29501 \
downstream_phase/run_phase_training.py \
--batch_size 12 \
--epochs 50 \
--save_ckpt_freq 10 \
--model surgformer_HTA_KCA \
--pretrained_path pretrain_params/TimeSformer_divST_8x32_224_K400.pyth \
--mixup 0.8 \
--cutmix 1.0 \
--smoothing 0.1 \
--lr 5e-4 \
--layer_decay 0.75 \
--warmup_epochs 5 \
--data_path /home/hinojoc/projects/2024/IBEX_TripletSurgRec/Surgformer/data/Cholec80 \
--eval_data_path /home/hinojoc/projects/2024/IBEX_TripletSurgRec/Surgformer/data/Cholec80 \
--nb_classes 7 \
--data_strategy online \
--output_mode key_frame \
--num_frames 16 \
--sampling_rate 4 \
--data_set Cholec80 \
--data_fps 1fps \
--output_dir /home/hinojoc/projects/2024/IBEX_TripletSurgRec/Surgformer/results/Cholec80_reprod_exp2 \
--log_dir /home/hinojoc/projects/2024/IBEX_TripletSurgRec/Surgformer/results/Cholec80_reprod_exp2 \
--num_workers 8 \
--dist_eval \
--cut_black \
```

- Did you use an effective batch size of 24, or was the effective batch size 24*3(GPUS)= 72 
- How long did the training take in your case?
- Can you please share the TimeSformer pretrained weights you used to initialize your model?

Thank you very much in advance!

**Comment by isyangshu** [comment]
@carlosh93 
Hi, would you mind sharing your final performance, and whether it is significantly different from what I reported? 
If possible, you also need to print each image path used for training to see if the training is done correctly, such as whether the frame sampling rate is wrong (1.png, 5.png or 0.png 25.png, because I changed the image naming method once), whether it overflows across the video boundary.

I use the batch size 24 for all experiments. I need to train for more than 40 hours, and the best results appear in 10 to 20 epochs. Meanwhile, the actual optimal performance is obtained by using 24*4 for testing.

Here is the pretrained parameters I used for initialization:
https://hkustconnect-my.sharepoint.com/:u:/g/personal/syangcw_connect_ust_hk/EYPZD_PyU5JChTUatgsw0QgBGwdsfFoLcuU6fTdTgiCD7A?e=uM5KO8

**Comment by carlosh93** [comment]
Sure @isyangshu, here are my logs, I achieved 82.09% accuracy. This run used a batch size of 8 per GPU, so the effective batch size is 16. Right now, I'm training with a batch size of 12 per GPU, but the results do not seem to improve.

Also, I just noted that I got some nan in the  grad_norm during training. Did this happen to you? I think I'm using the same dataset and the pkl files you provided [here](https://hkustconnect-my.sharepoint.com/:f:/g/personal/syangcw_connect_ust_hk/Et8_FZe9ahdKkSWNe98ygWsBiBAKdE1iarDyn4oeaZ26Pg?e=ycZxtM). By the way, it would also be helpful if you please share your logs so I can do an in-depth comparison.

Thank you!

[Archive.zip](https://github.com/user-attachments/files/19834147/Archive.zip)

**Comment by isyangshu** [comment]
Your result is also significantly lower than other people's reproductions from this issue.
I'll spend some time trying to figure out the problem.
However, due to my negligence when merging the two versions of data, I suggest you review the data problem.

**Comment by carlosh93** [comment]
No problem, I will check the data, maybe that's why I'm getting those nans. If you find your training logs, please share them here so I can also do an in-depth comparison between losses, grad_norm, etc. Finally, just to confirm, did you set in the train_phase.sh `--batch_size 8` right ? so you are using a batch size of 8 per gpu (so effective batch size is 8*3=24).

**Comment by isyangshu** [comment]
Yes, I used batch size 8.
If you don't use deepspeed, the NAN in grad_norm is normal.

Here is my log for Cholec80 with 16*4:
[log.txt](https://github.com/user-attachments/files/19834331/log.txt)
